builtin/logical/pki/path_tidy.go

```diff
@@ -1038,152 +1085,163 @@ func (b *backend) doTidyRevocationStore(ctx context.Context, req *logical.Reques
 		return false, err
 	}
 
-	rebuildCRL := false
-
-	revokedSerials, err := req.Storage.List(ctx, "revoked/")
-	if err != nil {
-		return false, fmt.Errorf("error fetching list of revoked certs: %w", err)
-	}
-
+	var after string
+	var revokedSerialsCount uint
 	haveWarned := false
-	revokedSerialsCount := len(revokedSerials) + int(revokedDeleted)
-	metrics.SetGauge([]string{"secrets", "pki", "tidy", "revoked_cert_total_entries"}, float32(revokedSerialsCount))
-
+	rebuildCRL := false
 	fixedIssuers := 0
 
-	revokedSafetyBuffer := config.SafetyBuffer
-	if config.RevokedSafetyBuffer != nil {
-		revokedSafetyBuffer = *config.RevokedSafetyBuffer
-	}
-
-	var revInfo revocationInfo
-	for i, serial := range revokedSerials {
-		b.tidyStatusMessage(fmt.Sprintf("Tidying revoked certificates: checking certificate %d of %d", i, len(revokedSerials)))
-		metrics.SetGauge([]string{"secrets", "pki", "tidy", "revoked_cert_current_entry"}, float32(i))
-
-		// Check for cancel before continuing.
-		if atomic.CompareAndSwapUint32(b.tidyCancelCAS, 1, 0) {
-			return false, tidyCancelledError
+	for {
+		revokedSerials, err := req.Storage.ListPage(ctx, "revoked/", after, config.PageSize)
+		if err != nil {
+			return false, fmt.Errorf("error fetching list of revoked certs: %w", err)
 		}
 
-		// Check for pause duration to reduce resource consumption.
-		if config.PauseDuration > (0 * time.Second) {
-			b.revokeStorageLock.Unlock()
-			time.Sleep(config.PauseDuration)
-			b.revokeStorageLock.Lock()
+		// If no revokedSerials are returned, we've reached the end of the list
+		if len(revokedSerials) == 0 {
+			break
 		}
 
-		revokedEntry, err := req.Storage.Get(ctx, "revoked/"+serial)
-		if err != nil {
-			return false, fmt.Errorf("unable to fetch revoked cert with serial %q: %w", serial, err)
-		}
+		revokedSerialsCount += uint(len(revokedSerials))
+		metrics.SetGauge([]string{"secrets", "pki", "tidy", "revoked_cert_total_entries"}, float32(revokedSerialsCount))
```

Comment:
I think you did this correctly for `doTidyCertStore` but I think this should only be done once, at the end of the function.